### PR TITLE
Added missing docs for CKEDITOR.htmlParser.filterRulesDefinition

### LINKED ISSUE #2387: filterRulesDefinition docs are empty
## Type of report

Task

## Provide description of the task

[`filterRulesDefinition`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_htmlParser_filterRulesDefinition.html) docs are empty. We need to document its members.

### PR DESCRIPTION
**Pull request by engineering-this**
## What is the purpose of this pull request?

Task

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [ ] Unit tests
- [ ] Manual tests

## What changes did you make?

Added missing docs for CKEDITOR.htmlParser.filterRulesDefinition and its properties.

Closes #2387

### CODE REVIEW

File: core/htmlparser/filter.js [hunk 1]
```diff
@@ -402,6 +402,36 @@
 } )();
 
 /**
+ * This is an abstract class representing filter rules passed to {@link CKEDITOR.htmlParser.filter#constructor} as an argument. See {@link CKEDITOR.htmlParser.filter} for examples.
```
Comment (jacekbogdanski) [inline core/htmlparser/filter.js]: I'm not sure if limiting this class to filter constructor is a good idea. This class may be used in different places - moreover, it's already used as an argument for [addRules](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_htmlParser_filter.html#method-addRules) function. Also in the future, we may want to use it in other places which with the proposed documentation shape will require another refactoring.

Also, mentioning that this class is abstract and used in some other places is redundant. This information is already viewable from functions arguments and returned types.

Instead of linking `htmlParser.filter` for examples, I think that it would be nice to attach some example code of definition and refactor duplicated code from filter listing.

Comment (jacekbogdanski) [reply core/htmlparser/filter.js]: I would extend this description with the information like:

What this class represents?
Why would I have to use it?
How I have to use it?

not implementation details.

File: core/htmlparser/filter.js [hunk 2]
```diff
@@ -402,6 +402,140 @@
...
+ * Examples:
+ *
+ * ```javascript
+ * elementNames: [
+ * 		[ /^div$/, 'p' ], // Converts 'div' into 'p'.
+ * 		[ /^cke:?/, '' ] // Removes 'cke:' prefixes.
+ * ]
+ * ```
+ *
+ * Incorrect:
```
Comment (jacekbogdanski) [inline core/htmlparser/filter.js]: Examples already describe correct rule regex, not sure if we need this section. It seems obvious to me that if you are matching string by regex, too small specificity will give you invalid results. WDYT?

File: core/htmlparser/filter.js [hunk 3]
```diff
@@ -402,6 +402,140 @@
...
+ * 		[ 'p', 'div' ] // Converts every letter 'p' inside element name into 'div'
+ * 		// 'p' element will be converted into 'div', 'span' wil be converted into 'sdivan'.
+ * 		// To convert only 'p' elements use regex `/^p$/`.
+ * ]
+ * ```
+ *
+ */
+
+/**
+ * @property {String[][]/RegExp[][]} attributeNames An array of rules for data processing attribute names.
```
Comment (jacekbogdanski) [inline core/htmlparser/filter.js]: The same as above.

File: core/htmlparser/filter.js [hunk 4]
```diff
@@ -402,6 +402,140 @@
...
+ * 		[ /^data-custom$/, 'data-cke' ]
+ * 		// Converts attribute 'data-custom' into 'data-cke'.
+ * ]
+ * ```
+ *
+ */
+
+/**
+ * @property {Object.<String, Function>} elements An object containing pairs of keys representing
+ * element selectors and function for data processing selected elements. If function contains return statement element will be removed.
```
Comment (jacekbogdanski) [inline core/htmlparser/filter.js]: I would split this long comment into two lines (after the dot).

Also, sencence `If function contains return statement element will be removed` is invalid. Every javascript function returns something (also without a return statement, which is `undefined`) so you should be more precise about return type. What if I write:
```javascript
function() {
  return undefined;
}
```
Will element still be removed?

Comment (jacekbogdanski) [reply core/htmlparser/filter.js]: This `data processing` phrase is very mysterious to me. I really don't understand what it does and what my function should do. Something like `transformation function` is more understable for me than it, although it may be subjective. E.g.

`An object containing pairs of keys representing element selectors and transformation function.`

Also, is the element really selected inside the editor? What if I apply filter into document fragment?

File: core/htmlparser/filter.js [hunk 5]
```diff
@@ -402,6 +402,140 @@
...
+ *
+ * ```javascript
+ * attributes: {
+ * 		'class': function( value, element ) {
+ * 			if ( element.name === 'div' ) {
+ * 				return value + ' cke_div' // Adds class 'cke_div' to every div element.
+ * 			}
+ * 		},
+ * 		id: function() {
+ * 			return false; // Removes every elements 'id' attribute.
```
Comment (jacekbogdanski) [inline core/htmlparser/filter.js]: I would rephrase it into `Removes processed element id attribute`. `Every` word is so wide that it may introduce some misconceptions. It applies to the whole proposed documentation.

File: core/htmlparser/filter.js [hunk 6]
```diff
@@ -402,6 +402,140 @@
...
+ */
+
+/**
+ * @property {Function} root Function for data processing root element.
+ *
+ * Examples:
+ *
+ * ```javascript
+ * root: function( element ) {
+ * 		element.children.push( someElement ) // Appends child to root element.
```
Comment (jacekbogdanski) [inline core/htmlparser/filter.js]: Missing semicolon.

File: core/htmlparser/filter.js [hunk 7]
```diff
@@ -402,6 +402,140 @@
...
+ * to be applied for transforming and filtering content upon data processing.
+ *
+ * Filter rules definition can be also added with {@link CKEDITOR.htmlParser.filter.addrules}.
+ *
  * @class CKEDITOR.htmlParser.filterRulesDefinition
  * @abstract
  */
+
+/**
+ * @property {String[][]/RegExp[][]} elementNames An array of rules for data processing element names.
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply core/htmlparser/filter.js]: I'm sure that a new type would require some discussion on our team.

File: core/htmlparser/filter.js [hunk 8]
```diff
@@ -402,6 +402,144 @@
...
+ * and functions used upon element filtering and transformation.
+ *
+ * A selector can be either element name or one of following: '^', '$'.
+ *
+ * '^' and '$' is to be applied on every filtered element. The first is applied before element specific filter,
+ * and second is applied after element specific filter.
+ *
+ * Function can contain return statement:
+ * - If `false` is returned the element is removed.
+ * - If an other element is returned it overwrites the original element.
```
Comment (jacekbogdanski) [inline core/htmlparser/filter.js]: Actually, returning element always overwrites the original element, although new type element won't be processed anymore: https://github.com/ckeditor/ckeditor-dev/blob/master/core/htmlparser/filter.js#L347
What do you think about:

> Additionaly function can contain return statement:                                                                                                                                                                                    
> - `false` will remove whole element from the node tree.                                                                                                                                                                               
> - {@link CKEDITOR.dom.element element} will overwrite the original element.

File: core/htmlparser/filter.js [hunk 9]
```diff
@@ -402,6 +402,144 @@
...
+ *
+ * Definition object represents rules as a set of properties with callback functions
+ * to be applied for transforming and filtering content upon data processing.
+ *
  * @class CKEDITOR.htmlParser.filterRulesDefinition
  * @abstract
  */
+
+/**
+ * @property {Array<String,RegExp>[]} elementNames An array of rules for element names transformation.
```
Comment (jacekbogdanski) [inline core/htmlparser/filter.js]: It seems like umberto/jsduck incorrectly parsers this construction :thinking:

File: core/htmlparser/filter.js [hunk 10]
```diff
@@ -402,6 +402,164 @@
...
+ *
+ * Definition object represents rules as a set of properties with callback functions
+ * to be applied for transforming and filtering content upon data processing.
+ *
  * @class CKEDITOR.htmlParser.filterRulesDefinition
  * @abstract
  */
+
+/**
+ * @property {CKEDITOR.htmlParser.nameFilterRule[]} elementNames An array of rules for element names transformation.
```
Comment (Comandeer) [inline core/htmlparser/filter.js]: TBH I do not like how this custom type looks like. I would think about replacing it with `Array.<RexExp,String>` syntax (note the dot that magically fixes jsduck issues; it would be also consistent with `Object.<whatever>` syntax that we use elsewhere).

Comment (engineering-this) [reply core/htmlparser/filter.js]: Yes it was my mistake, to skip the dot, yet it is an array of arrays, which is not correctly parsed by jsduck.
`Array.<RegExp,String>[]` -> `Array.&ltRegExp,String>[]`

File: core/htmlparser/nameFilterRule.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+/**
+ * Abstract class describing the definition of {@link CKEDITOR.htmlParser.filterRulesDefinition} `elementNames` and `attributesNames` filtering rules.
+ *
+ * ```javascript
+ *  var rule = [ /^div$/, 'p' ];
+ * ```
+ *
+ * @class CKEDITOR.htmlParser.nameFilterRule
```
Comment (Comandeer) [inline core/htmlparser/nameFilterRule.js]: I'm wondering about the name of the class. It's used for elements's transformations, so maybe `CKEDITOR.htmlParse.transformRule`?

Comment (engineering-this) [reply core/htmlparser/nameFilterRule.js]: Every member of `CKEDITOR.htmlParser.filterRulesDefinition` class can used for some kind of transformation. Although only two uses this abstract class. `transformRule` sounds a bit misleading to me.
What do you think about `nameTransformRule`? These rules alters element names or attributes, but they don't touch values/content so this name would be a bit more precise. WDYT @Comandeer?

**Comment by engineering-this** [comment]
- Added links to methods that uses `filterRulesDefinition`.
- Renamed `nameFilterRule` to `nameTransformRule`.
